compiler/plc_xml/src/model/fbd.rs

```diff
@@ -105,14 +122,239 @@ impl Parseable for FunctionBlockDiagram {
         }
 
         nodes.sort_by(|_, b, _, d| b.partial_cmp(d).unwrap()); // This _shouldn't_ panic because our `partial_cmp` method covers all cases
+
         Ok(FunctionBlockDiagram { nodes })
     }
 }
 
+enum SourceReference {
+    Value(NodeId),
+    Connector(NodeId),
+    Unconnected(NodeId),
+}
+
+struct ResolvedConnection {
+    id: NodeId,
+    resolved_ref_id: NodeId,
+    block_parameter_index: Option<usize>,
+}
+
+// IndexMap<NodeId, Node> interface for connection-point (sink/source) desugaring
+trait ConnectionResolver<'xml> {
+    fn desugar_connection_points(
+        &mut self,
+        source_location_factory: &SourceLocationFactory,
+    ) -> Result<(), Vec<Diagnostic>>;
+    fn get_source_references(&self) -> HashMap<&str, SourceReference>;
+    fn get_resolved_connection_ids(
+        &self,
+        source_connections: &HashMap<&str, SourceReference>,
+        source_location_factory: &SourceLocationFactory,
+    ) -> (Vec<ResolvedConnection>, Vec<Diagnostic>);
+    fn find_assignable_sink_value(
+        &self,
+        source_connections: &HashMap<&str, SourceReference>,
+        connector_id: NodeId,
+        connector_path: &mut IndexSet<String>,
+        source_location_factory: &SourceLocationFactory,
+    ) -> Result<NodeId, Diagnostic>;
+}
+
+impl<'xml> ConnectionResolver<'xml> for NodeIndex<'xml> {
+    /// Updates all nodes in the index, which are connected via connection-points (sink/source) to be treated as
+    /// if they are connected directly instead.
+    ///
+    /// ```st
+    /// // assignments using sink and source
+    /// INPUT ━━━━> SOURCE ┅┅┅> SINK ━┳━━> OUT1
+    ///                               ┣━━> OUT2
+    ///                               ┗━━> OUT3
+    /// // resolve to
+    /// INPUT ━┳━━> OUT1
+    ///        ┣━━> OUT2
+    ///        ┗━━> OUT3
+    /// ```
+    fn desugar_connection_points(
+        &mut self,
+        source_location_factory: &SourceLocationFactory,
+    ) -> Result<(), Vec<Diagnostic>> {
+        let source_connections = self.get_source_references();
+
+        let (resolved_connections, diagnostics) =
+            self.get_resolved_connection_ids(&source_connections, source_location_factory);
+
+        // update sink-connections to point to assignable values of associated source
+        for ResolvedConnection { id, resolved_ref_id, block_parameter_index } in resolved_connections {
+            self.entry(id).and_modify(|node| match node {
+                Node::Block(block) => {
+                    block.variables.get_mut(block_parameter_index.unwrap()).unwrap().ref_local_id =
```

Comment:
Could we try to avoid the double unwraps? We return Results anyway here.

Reply:
Since all the `ResolvedConnection` structs in `resolved_connections` are only created for cases with valid values, I felt the double unwrap here was fine. But I'm happy to change these into results anyway